sample/weighted.go

```diff
@@ -0,0 +1,118 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import (
+	"math/rand"
+
+	"github.com/gonum/floats"
+)
+
+// Weighted provides sampling without replacement from a collection of items with
+// non-uniform probability.
+type Weighted struct {
+	weights []float64
+	// heap is a weight heap. See comments
+	// in contained/heap for an explanation.
+	heap []float64
+	src  *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the weights w. If src is nil, rand.Rand is
+// used as the random source.
+func NewWeighted(w []float64, src *rand.Rand) Weighted {
+	s := Weighted{
+		weights: make([]float64, len(w)),
+		heap:    make([]float64, len(w)),
+	}
+	s.ReweightAll(w)
+	return s
+}
+
+// Len returns the number of items held by the Weighted, including items
+// already taken.
+func (s Weighted) Len() int { return len(s.weights) }
+
+// Take returns an index from the Weighted with probability proportional
+// to the weight of the item. The weight of the item is then set to zero.
+// Take returns false if there are no items remaining.
+func (s Weighted) Take() (idx int, ok bool) {
+	if floats.EqualWithinAbsOrRel(s.heap[0], 0, 1e-12, 1e-12) {
+		return -1, false
+	}
+
+	var r float64
+	if s.src == nil {
+		r = s.heap[0] * rand.Float64()
+	} else {
+		r = s.heap[0] * s.src.Float64()
+	}
+	i := 1
+	last := -1
+	for {
+		if r -= s.weights[i-1]; r <= 0 {
+			break // fall within item i-1
+		}
+		i <<= 1 // move to left child
+		if d := s.heap[i-1]; r > d {
+			r -= d
+			// if enough r to pass left child
+			// move to right child
+			// state will be caught at break above
+			i++
+		}
+		if i == last {
+			// No progression.
```

Comment:
What does this case mean? I'm surprised this doesn't result it returning an error.